### ISSUE TITLE
a little feature request for embark-indicator

### ISSUE THREAD

**Issue by jaor** [oantolin/embark#132]
i was writing my own indicator function and it'd be nice to have
available there, besides the keymap, the target (to show it together
with the keymap description)... afaict, that could be done
[here](https://github.com/oantolin/embark/blob/master/embark.el#L561),
but it'd be an incompatible change.  or maybe there's already a way of
accessing the current target that i've missed?

thanks!

**Comment by oantolin** [comment]
This is a good idea. It can be done backwards compatibly changing that line to:

```elisp
(condition-case nil
    (funcall indicator keymap target)
  (wrong-number-of-arguments (funcall indicator keymap)))
```

But maybe I should just change it to `(funcall indicator keymap target)`. The string case can also be changed to be a format string that gets used as `(format indicator target)`.

**Comment by minad** [comment]
@jaor What kind of indicator function are you writing? Can you please post it somewhere since it may be interesting for a wider audience? I guess you will do so on your blog?

Maybe you have seen #50 which improves the which-key menu, in case you are using that. There is also an open PR for menu-items in which-key, which makes it easier to provide a user-friendly menu out of the box.

**Comment by jaor** [comment]
On Tue, Jan 26 2021, Daniel Mendler wrote:

> @jaor What kind of indicator function are you writing? Can you please
> post it somewhere since it may be interesting for a wider audience? I
> guess you will do so on your blog? 😄

eventually :)  a very simple one that i quite like because it's all
built-in was:

```elisp
(defun jao-embark-indicator (keymap)
  (let ((display-buffer-alist '(("*Help*"
                                 (display-buffer-at-bottom)
                                 (window-parameters (mode-line-format . none))
                                 (window-height . fit-window-to-buffer))))
        (inhibit-message t))
    (describe-keymap keymap)
    (lambda () (kill-buffer-and-window "*Help*"))))
```

it has the (little?) drawback of messing with the user's *Help*
buffer... i don't mind because i use helpful, but otherwise perhaps is
better to not kill it.  or possibly tell describe-keymap to use a
different buffer, i haven't investigated that.

i'm actually using a more convoluted one, which i'm sure can be written
more succintly (the kinda messy reduce with jao-embark--bind-desc is
used to discover column lengths and print them--key, command,
docstring--aligned... i'm sure emacs has something to print columns
quickly, but i don't know it):

https://jao.io/cgit/emacs/tree/consult.org#n113

50 loc is probably too long for an indicator, but it looks pretty :)

**Comment by jaor** [comment]
> But maybe I should just change it to (funcall indicator keymap
> target).

yeah, simpler and you have the "breaking changes" to warn... i guess not
many people has custom indicators yet.

> The string case can also be changed to be a format string that gets
> used as (format indicator target).

i think that's a very good idea.

**Comment by oantolin** [comment]
> yeah, simpler and you have the "breaking changes" to warn... i guess not many people has custom indicators yet.

Actually, probably a bunch of people use the which-key one from the wiki. I think I'll issue a warning for a while.

**Comment by jaor** [comment]
btw, another thing i'd like to have is something similar to the
embark-keymap-help pseudo-command, but to invoke my embark indicator on
demand, as it's done now with embark-completing-read-prompter after
C-h... i would then set embark-indicator to its default, and then hook
the invocation of my own indicator to, say, C-h: that would invoke my
indicator and stay in embark-keymap-prompter, so that i don't need to
replicate the latter's logic in a full new prompter.

to put it another way, i'd like to be able to customize the function
called when the prompt receives the command embark-keymap-help.  but i
can always :override embark-completing-read-prompter with the one i
want, so maybe this is not worth complicating embark proper.

**Comment by oantolin** [comment]
It's easier than that: just bind your help command to `C-h` (or whatever key you wish) in `embark-meta-map`.

**Comment by jaor** [comment]
On Tue, Jan 26 2021, Omar Antolín Camarena wrote:

> It's easier than that: just bind your help command to C-h (or whatever
> key you wish) in embark-meta-map.

i'm probably missing something here: how would i have access to the
active keymap and action in that help command? i want the special
treatment the original C-h receives in embark-keymap-prompter...

**Comment by oantolin** [comment]
You're not missing anything, I got confused.

**Comment by jaor** [comment]
oh, a dirty trick just crossed my mind: couldn't i in my
embark-indicator function (which will receive keymap and action) simply
display the regular "Action for ..." message but add to the keymap it
receives the binding for C-h showing extended help?

**Comment by oantolin** [comment]
I made the change you requested for indicator functions and now treat indicator strings as format strings. Please test, and please read the docstrings (rewriting them was by far the hardest part of this 😛). I made your request about `C-h` into a separate issue, so I don't forget about it.

**Comment by jaor** [comment]
On Tue, Jan 26 2021, Omar Antolín Camarena wrote:

> I made the change you requested for indicator functions and now treat
> indicator strings as format strings.  Please test, and please read the
> docstrings (rewriting them was by far the hardest part of this 😛).

it's working very well for me, thanks! (a nice touch is that the target
is received with any text properties in has in its buffer, so building
an indicator has nice results with little effort... this made me think
of the possibility of meta-targets that store a target in their
properties (or the other way round), but i digress! :)

the docstrings also look good to me; even though i'm probably not the
best positioned to judge, they seem very clear to me.

>  I made your request about C-h into a separate issue, so I don't
> forget about it.

excellent.  i'm so happy with my new embark-indicator that i'm not sure
i'll want to hide it in the end! :) (just kidding: i'm sure that i'll
eventually know by heart the most used actions keys and won't even look
at it).
